[dtensor][6/n] refactor reduction to use op strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109145
* #109144
* __->__ #108264
* #108263

This PR refactors the reduction op to use strategy based propagation

wanchaol committed:
[dtensor][6/n] refactor reduction to use op strategy

This PR refactors the reduction op to use strategy based propagation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][6/n] refactor reduction to use op strategy"


This PR refactors the reduction op to use strategy based propagation

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][6/n] refactor reduction to use op strategy"


This PR refactors the reduction op to use strategy based propagation

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -1,54 +1,211 @@
+    # by default follow reduction input strategy
+    reduction_strategy = OpStrategy([])
+
+    for strat in input_strategy.strategies:
```
nit: I googled what is the best abbreviation of "strategy", looks like people like "[STRTG](https://www.abbreviations.com/term/2084962/strategy)" better..

fduwjj reviewed (approved):
LGTM

wanchaol committed:
Update on "[dtensor][6/n] refactor reduction to use op strategy"


This PR refactors the reduction op to use strategy based propagation

[ghstack-poisoned]
